chain-impl-mockchain/src/leadership/genesis/vrfeval.rs

```diff
@@ -78,55 +102,71 @@ impl Input {
 pub type Witness = <Curve25519_2HashDH as VerifiableRandomFunction>::VerifiedRandomOutput;
 pub type WitnessOutput = <Curve25519_2HashDH as VerifiableRandomFunction>::RandomOutput;
 
-/// Evaluate if the threshold is above for a given input for the key and the associated stake
-///
-/// On threshold success, the witness is returned, otherwise None is returned
-pub fn evaluate(
-    my_stake: PercentStake,
-    key: &SecretKey<Curve25519_2HashDH>,
-    nonce: &Nonce,
-    slotid: SlotId,
-) -> Option<Witness> {
-    let input = Input::create(nonce, slotid);
-    let csprng = OsRng::new().unwrap();
-    let vr = vrf_evaluate_and_proove(key, &input.0, csprng);
-    let r = vrf_verified_get_output::<Curve25519_2HashDH>(&vr);
-    let t = get_threshold(&input, &r);
-    if above_stake_threshold(t, my_stake) {
-        Some(vr)
-    } else {
-        None
-    }
+pub struct Evaluator<'a> {
+    pub my_stake: PercentStake,
+    pub key: &'a SecretKey<Curve25519_2HashDH>,
+    pub nonce: &'a Nonce,
+    pub slot_id: SlotId,
+    pub param_f: F,
 }
 
-/// verify that the witness pass the threshold for this witness for a given
-/// key and its associated stake.
-///
-/// On success, the nonce is returned, otherwise None is returned
-pub fn verify(
-    key_stake: PercentStake,
-    key: &PublicKey<Curve25519_2HashDH>,
-    nonce: &Nonce,
-    slotid: SlotId,
-    witness: &Witness,
-) -> Option<Nonce> {
-    let input = Input::create(nonce, slotid);
-    if vrf_verify(key, &input.0, witness) == VRFVerification::Success {
-        let r = vrf_verified_get_output::<Curve25519_2HashDH>(witness);
+impl<'a> Evaluator<'a> {
+    /// Evaluate if the threshold is above for a given input for the key and the associated stake
+    ///
+    /// On threshold success, the witness is returned, otherwise None is returned
+    pub fn evaluate(&self) -> Option<Witness> {
+        let input = Input::create(self.nonce, self.slot_id);
+        let csprng = OsRng::new().unwrap();
+        let vr = vrf_evaluate_and_proove(self.key, &input.0, csprng);
+        let r = vrf_verified_get_output::<Curve25519_2HashDH>(&vr);
         let t = get_threshold(&input, &r);
-        if above_stake_threshold(t, key_stake) {
-            Some(get_nonce(&input, &r))
+        if above_stake_threshold(t, &self.my_stake, self.param_f) {
+            Some(vr)
         } else {
             None
         }
-    } else {
-        None
     }
 }
 
-fn above_stake_threshold(threshold: Threshold, stake: PercentStake) -> bool {
-    // TODO F is hardcoded here
-    threshold >= phi(F::create(0.5), stake)
+pub struct Verifier<'a> {
+    pub key_stake: PercentStake,
+    pub key: &'a PublicKey<Curve25519_2HashDH>,
+    pub nonce: &'a Nonce,
+    pub slot_id: SlotId,
+    pub witness: &'a Witness,
+    pub param_f: F,
```

Comment:
I wonder if common parts of `Evaluator` and `Verifier` could be made into a single type.